daemon/logger/jsonfilelog/read.go

```diff
@@ -121,96 +119,106 @@ func tailFile(f io.ReadSeeker, logWatcher *logger.LogWatcher, tail int, since ti
 	}
 }
 
-func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
-	dec := json.NewDecoder(f)
-	l := &jsonlog.JSONLog{}
+func (l *JSONFileLogger) followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, since time.Time) {
+	var (
+		msg        *logger.Message
+		err        error
+		rotated    bool
+		readIndex  uint64
+		writeIndex uint64
+
+		dec        = json.NewDecoder(f)
+		log        = &jsonlog.JSONLog{}
+		subscriber = l.writeNotifier.Subscribe()
+		watchClose = logWatcher.WatchClose()
+	)
+
+	go func() {
+		for {
+			if _, ok := <-subscriber; !ok {
```

Comment:
@cpuguy83 Here is a demo about how the message will be lost if we use the subscriber's channel directly. https://gist.github.com/mountkin/8d982e4611467ba87ac2
That's why I didn't do it this way
```go
select {
	case <-watchClose:
		// stuff
	case <-notifyRotate:
		// stuff
	case <-subscriber:
		// stuff
		logWatcher.Msg <- msg  // This may block if the network connection is slow. see https://github.com/docker/docker/blob/master/daemon/logs.go#L84
}
```